### ISSUE TITLE
Classpath issues in 1.2.0

### ISSUE THREAD

**Issue by hrach** [cashapp/paparazzi#918]
**Description**
After updating to Paparazzi 1.3.0 (through 1.2.0), we have multiple issues running the unit test/mockk/android lint, probably caused by some classpath issues. 

**Steps to Reproduce**

Running unit tests fail when Mockk mocks platforms types. E.g.
calling 

```
mockk<android.telephony.TelephonyManager>()
``` 
throws:

```
Could not initialize class android.telephony.TelephonyManager
java.lang.NoClassDefFoundError: Could not initialize class android.telephony.TelephonyManager
	at jdk.internal.reflect.GeneratedSerializationConstructorAccessor14.newInstance(Unknown Source)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:483)
	at org.objenesis.instantiator.sun.SunReflectionFactoryInstantiator.newInstance(SunReflectionFactoryInstantiator.java:48)
	at io.mockk.proxy.jvm.ObjenesisInstantiator.instanceViaObjenesis(ObjenesisInstantiator.kt:76)
	at io.mockk.proxy.jvm.ObjenesisInstantiator.instantiateViaProxy(ObjenesisInstantiator.kt:67)
	at io.mockk.proxy.jvm.ObjenesisInstantiator.instance(ObjenesisInstantiator.kt:30)
	at io.mockk.proxy.common.ProxyMaker.instantiate(ProxyMaker.kt:91)
	at io.mockk.proxy.common.ProxyMaker.proxy(ProxyMaker.kt:60)
	at io.mockk.impl.instantiation.JvmMockFactory.newProxy(JvmMockFactory.kt:34)
	at io.mockk.impl.instantiation.AbstractMockFactory.newProxy$default(AbstractMockFactory.kt:24)
	at io.mockk.impl.instantiation.AbstractMockFactory.mockk(AbstractMockFactory.kt:59)
```

----

Running Android lint on our project ends up analyzing dependencies:
```
/builds/mobile/android/cache/.gradle/caches/modules-2/files-2.1/org.bouncycastle/bcpkix-jdk15on/1.67/5f48020a2a60a8d6bcbecceca23529d225b28efb/bcpkix-jdk15on-1.67.jar: 
Error: checkClientTrusted is empty, which could cause insecure network traffic due to trusting arbitrary TLS/SSL certificates presented by peers [TrustAllX509TrustManager]
``` 

----

Running Android Tests ends up with conflicts in classpath:

```
Execution failed for task ':app:checkProductionGoogleInstrumentationAndroidTestDuplicateClasses'.
> A failure occurred while executing com.android.build.gradle.internal.tasks.CheckDuplicatesRunnable
   > Duplicate class com.google.android.apps.common.testing.accessibility.framework.AccessibilityCheck found in modules accessibility-test-framework-3.1.2-runtime (com.google.android.apps.common.testing.accessibility.framework:accessibility-test-framework:3.1.2) and layoutlib-native-jdk11-2022.2.1-5128371-2 (app.cash.paparazzi:layoutlib-native-jdk11:2022.2.1-5128371-2)
```

All those issues are triggered exclusively by Paparazzi update.

**Additional information:**
- Paparazzi Version: 1.3.0
- OS: any
- Compile SDK: 33
- Gradle Version: 8.1.1
- Android Gradle Plugin Version: 8.0.1

**Comment by hrach** [comment]
Similar/same report here: https://kotlinlang.slack.com/archives/C5HT9AL7Q/p1669397250324199

**Comment by jrodbx** [comment]
At first glance, this looks like 3 different bugs, all unrelated to Paparazzi.  The second one confuses me the most, would need an example branch to dig in more.

Re: the first one, since Layoutlib itself mocks out Android platform types, any attempt to mock them directly in tests would fail.  I don't know if there's any good long-term solution around this.  Open to suggestions though, if I'm missing something and someone else in the community has ideas.

Additionally, Paparazzi aims somewhat to force consumers to design their views/composables to be as "simple" as possible.  Is it possible to push TelephonyManager out of your view logic?  That would also solve the issue and be the preferred way here.

Re: the third one, Paparazzi isn't meant to be run in instrumentation tests, i.e., androidTest.  This is because androidTest source set runs on device/emulator and layoutlib is meant to mock Android out on the jvm.  I suspect you have a Gradle misconfiguration and/or added Paparazzi tests into /androidTest, both of which would be user error.  Am I mistaken?

Admittedly, without a sample/example branch, most of what I'm explaining is based on guesses.  Do you have a minimal example I can dig into more?

**Comment by hrach** [comment]
Thanks for the quick response!

We don't run paparazzi in Android tests nor do we use TelephonyManager in UI, it just sits in the same module that has this in domain logic. 

But you helped me to realize that there will be a bug - during the update I introduced shared BaseScreenshotTest into our shared testing module, that's how the Paparazzi classes got into AndroidTests which reuse this testing module as well.

I'll fix this and will get back with some minimal repro if possible. Thx.

**Comment by hrach** [comment]
So I solved all issues by:
- using a shared testing module with `implementation(libs.paparazzi)` only in unit test, not androidTests.
- not running screenshot tests during unit tests and not running unit tests during screenshots tests.

To do so, I've done something like:

```kotlin
tasks.withType<PaparazziPlugin.PaparazziTask> {
  val isPaparazzi = gradle.startParameter.taskNames.any { it.contains("Paparazzi") }
  if (isPaparazzi) {
    setTestNameIncludePatterns(listOf("*ScreenshotTest"))
  }
}
```

and 

```kotlin
testOptions {
  unitTests {
    all {
      val isPaparazzi = gradle.startParameter.taskNames.any { it.contains("Paparazzi") }
      if (!isPaparazzi) {
        it.exclude("**/*ScreenshotTest.class")
      }
    }
  }
}
```

Thanks for making leading me on the correct path.

**Comment by bilgehankalkan** [comment]
Hi,

I'm still having this issue even after applying [solution above](https://github.com/cashapp/paparazzi/issues/918#issuecomment-1578297382).

I tried applying `PaparazziPlugin` by checking `isPaparazzi` and that worked. But on our CI checks, we invoke `testDebugUnitTest` along with `verifyPaparazziDebug` task and causes to Paparazzi plugin to be applied and not triggers triggers normal unit tests due do `exclude` call.

I applied plugin like in [here](https://gist.github.com/bilgehankalkan/ee11ae626d968eb70ba2897420681981).

Is there any other solution or suggestion to fix this problem?
